presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java

```diff
@@ -100,6 +118,27 @@ public LookupSourceSupplierFactory compileLookupSourceFactory(List<? extends Typ
         return compileLookupSourceFactory(types, joinChannels, sortChannel, Optional.empty());
     }
 
+    public JoinCompiler()
+    {
+        this(createTestingIsDistinctFromGenerator());
+    }
+
+    private JoinCompiler(IsDistinctFromGenerator generator)
+    {
+        this.isDistinctFromGenerator = requireNonNull(generator, "generator is null");
+    }
+
+    private JoinCompiler(FunctionRegistry registry, TypeManager typeManager)
+    {
+        this.isDistinctFromGenerator = new IsDistinctFromGenerator(registry, typeManager);
+    }
+
+    @Inject
+    public JoinCompiler(Metadata metadata)
```

Comment:
You have 4 overloads here:

1. `public JoinCompiler()`
2. `private JoinCompiler(IsDistinctFromGenerator generator)`
3. `private JoinCompiler(FunctionRegistry registry, TypeManager typeManager)`
4. `public JoinCompiler(Metadata metadata)`

In Presto, we generally avoid having many overloads because doing so makes code messy and hard to maintain.

* Remove 1 and 2. For testing, you can use `MetadataManager.createTestMetadataManager()`
* For 3, it's ok to keep. It shows what are the actual parameters this object needs to initialize.
* For 4, it's ok to keep because we need it for Guice to work.